File: csharp/src/Signer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace TrueLayer.Signing
{
    /// <summary>
    /// Builder to generate a Tl-Signature header value.
    /// </summary>
    public abstract class Signer : Signer<Signer>
    {
        /// <summary>
        /// Start building a request Tl-Signature header value using private key
        /// RFC 7468 PEM-encoded data and the key's kid.
        /// </summary>
        public static Signer SignWithPem(string kid, ReadOnlySpan<char> privateKeyPem)
            => SignWith(kid, privateKeyPem.ParsePem());

        /// <summary>
        /// Start building a request Tl-Signature header value using private key
        /// RFC 7468 PEM-encoded data and the key's kid.
        /// </summary>
        public static Signer SignWithPem(string kid, ReadOnlySpan<byte> privateKeyPem)
#if (NETSTANDARD2_0)
            => SignWithPem(kid, Encoding.UTF8.GetString(privateKeyPem.ToArray()).AsSpan());
#else
            => SignWithPem(kid, Encoding.UTF8.GetString(privateKeyPem));
#endif

        /// <summary>
        /// Start building a request Tl-Signature header value using private key and the key's kid.
        /// </summary>
        public static Signer SignWith(string kid, ECDsa privateKey) => new PrivateKeySigner(kid, privateKey);

        /// <summary>
        /// Start building a request Tl-Signature header value using the key ID of the signing key (kid)
        /// and a function that accepts the payload to sign and returns the signature in IEEE P1363 format.
        /// </summary>
        public static AsyncSigner SignWithFunction(string kid, Func<string, Task<string>> signAsync) => new FunctionSigner(kid, signAsync);

        /// <summary>
        /// Initializes a new instance of the Signer class with the specified key ID.
        /// </summary>
        protected internal Signer(string kid) : base(kid)
        {
        }

        /// <summary>Produce a JWS `Tl-Signature` v2 header value.</summary>
        public abstract string Sign();
    }

    /// <summary>
    /// Base class for signature builders with fluent API support.
    /// </summary>
    public abstract class Signer<TSigner> where TSigner : Signer<TSigner>
    {
        private readonly string _kid;
        private readonly TSigner _this;
        /// <summary>HTTP method for the request.</summary>
        protected internal string _method = "POST";
        /// <summary>Request path.</summary>
        protected internal string _path = "";
        /// <summary>Request headers to include in the signature.</summary>
        protected internal readonly Dictionary<string, byte[]> _headers = new(StringComparer.OrdinalIgnoreCase);
        /// <summary>Request body.</summary>
        protected internal byte[] _body = Array.Empty<byte>();
        /// <summary>JSON Web Key Set URL.</summary>
        protected internal string? _jku;

        /// <summary>
        /// Initializes a new instance of the Signer class with the specified key ID.
        /// </summary>
        protected internal Signer(string kid)
        {
            _kid = kid;
            _this = (TSigner) this;
        }

        /// <summary>Add the request method, defaults to `"POST"` if unspecified.</summary>
        public TSigner Method(string method)
        {
            _method = method;
            return _this;
        }

        /// <summary>
        /// Add the request absolute path starting with a leading `/` and without any trailing slashes.
        /// </summary>
        public TSigner Path(string path)
        {
            if (!path.StartsWith("/"))
            {
                throw new ArgumentException($"Invalid path \"{path}\" must start with '/'");
            }

            _path = path;
            return _this;
        }

        /// <summary>
        /// Add a header name and value.
        /// May be called multiple times to add multiple different headers.
        /// <br/>
        /// Warning: Only a single value per header name is supported.
        /// </summary>
        public TSigner Header(string name, byte[] value)
        {
            _headers.Add(name.Trim(), value);
            return _this;
        }

        /// <summary>
        /// Add a header name and value.
        /// May be called multiple times to add multiple different headers.
        /// <br/>
        /// Warning: Only a single value per header name is supported.
        /// </summary>
        public TSigner Header(string name, string value) => Header(name, value.ToUtf8());

        /// <summary>
        /// Appends multiple header names and values.
        /// <br/>
        /// Warning: Only a single value per header name is supported.
        /// </summary>
        public TSigner Headers(IEnumerable<KeyValuePair<string, string>> headers)
        {
            foreach (var entry in headers)
            {
                Header(entry.Key, entry.Value);
            }

            return _this;
        }

        /// <summary>
        /// Appends multiple header names and values.
        /// <br/>
        /// Warning: Only a single value per header name is supported.
        /// </summary>
        public TSigner Headers(IEnumerable<KeyValuePair<string, byte[]>> headers)
        {
            foreach (var entry in headers)
            {
                Header(entry.Key, entry.Value);
            }

            return _this;
        }

        /// <summary>
        /// Add the full request body.
        /// Note: This *must* be identical to what is sent with the request.
        /// </summary>
        public TSigner Body(byte[] body)
        {
            _body = body;
            return _this;
        }

        /// <summary>
        /// Add the full request body.
        /// Note: This *must* be identical to what is sent with the request.
        /// <br/>
        /// In this method it is assumed the body will be encoded using UTF-8.
        /// </summary>
        public TSigner Body(string body) => Body(body.ToUtf8());
        
        /// <summary>
        /// Sets the jku (JSON Web Key Set URL) in the JWS headers of the signature.
        /// </summary>
        public TSigner Jku(string jku)
        {
            _jku = jku;
            return _this;
        }

        /// <summary>
        /// Creates the JWS headers for the signature.
        /// </summary>
        protected internal Dictionary<string, object> CreateJwsHeaders()
        {
            var jwsHeaders = new Dictionary<string, object>
            {
                {JwsHeaders.Alg, "ES512"},
                {JwsHeaders.Kid, _kid},
                {JwsHeaders.TlVersion, "2"},
                {JwsHeaders.TlHeaders, string.Join(",", _headers.Select(h => h.Key))},
            };

            if (_jku is not null)
            {
                jwsHeaders.Add(JwsHeaders.Jku, _jku);
            }

            return jwsHeaders;
        }
    }
    
    /// <summary>
    /// Builder to generate a Tl-Signature header value.
    /// </summary>
    public abstract class AsyncSigner : Signer<AsyncSigner>
    {
        /// <summary>
        /// Initializes a new instance of the AsyncSigner class with the specified key ID.
        /// </summary>
        protected internal AsyncSigner(string kid) : base(kid)
        {
        }

        /// <summary>Produce a JWS `Tl-Signature` v2 header value.</summary>
        public abstract Task<string> SignAsync();
    }

    internal sealed class PrivateKeySigner : Signer
    {
        private readonly ECDsa _key;

        internal PrivateKeySigner(string kid, ECDsa privateKey) : base(kid)
        {
            _key = privateKey;
        }
        
        public override string Sign()
        {
            var jwsHeaders = CreateJwsHeaders();
#if NET5_0_OR_GREATER
            var serializedJwsHeaders = JsonSerializer.SerializeToUtf8Bytes(jwsHeaders, SigningJsonContext.Default.DictionaryStringObject);
#else
            var serializedJwsHeaders = JsonSerializer.SerializeToUtf8Bytes(jwsHeaders);
#endif
            var serializedJwsHeadersB64 = Base64Url.Encode(serializedJwsHeaders);

            var headerList = _headers.Select(e => (e.Key, e.Value));
            var signingPayload = Util.BuildV2SigningPayload(_method, _path, headerList, _body);
            var signingPayloadB64 = Base64Url.Encode(signingPayload);

            var signingMessage = $"{serializedJwsHeadersB64}.{signingPayloadB64}";
            var signingMessageBytes = Encoding.UTF8.GetBytes(signingMessage);

            // Compute SHA-512 hash (ES512 uses SHA-512)
#if NET5_0_OR_GREATER
            var hash = SHA512.HashData(signingMessageBytes);
#else
            byte[] hash;
            using (var sha512 = SHA512.Create())
            {
                hash = sha512.ComputeHash(signingMessageBytes);
            }
#endif

            // Sign the hash using ECDSA - signature will be in IEEE P1363 format (r||s)
            var signature = _key.SignHash(hash);
            var signatureB64 = Base64Url.Encode(signature);

            // Return detached JWS format: header..signature (empty payload)
            return $"{serializedJwsHeadersB64}..{signatureB64}";
        }
    }
    
    internal sealed class FunctionSigner : AsyncSigner
    {
        private readonly Func<string, Task<string>> _signAsync;
        
        internal FunctionSigner(string kid, Func<string, Task<string>> signAsync) : base(kid)
        {
            _signAsync = signAsync;
        }
        
        public override async Task<string> SignAsync()
        {
            var jwsHeaders = CreateJwsHeaders();
#if NET5_0_OR_GREATER
            var serializedJwsHeaders = JsonSerializer.SerializeToUtf8Bytes(jwsHeaders, SigningJsonContext.Default.DictionaryStringObject);
#else
            var serializedJwsHeaders = JsonSerializer.SerializeToUtf8Bytes(jwsHeaders);
#endif
            var serializedJwsHeadersB64 = Base64Url.Encode(serializedJwsHeaders);

            var headerList = _headers.Select(e => (e.Key, e.Value));
            var signingPayload = Util.BuildV2SigningPayload(_method, _path, headerList, _body);
            var signingPayloadB64 = Base64Url.Encode(signingPayload);

            var signingMessage = $"{serializedJwsHeadersB64}.{signingPayloadB64}";

            var signature = await _signAsync(signingMessage);

            return $"{serializedJwsHeadersB64}..{signature}";
        }
    }
}

File: java/lib/src/main/java/com/truelayer/signing/Signer.java
package com.truelayer.signing;

import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.JWSObject;
import com.nimbusds.jose.Payload;
import com.nimbusds.jose.crypto.ECDSASigner;
import com.nimbusds.jose.jwk.ECKey;

import java.nio.charset.StandardCharsets;
import java.security.interfaces.ECPrivateKey;
import java.util.LinkedHashMap;
import java.util.Map;

import static com.truelayer.signing.Utils.jwsHeaderMap;

/**
 * Builder to generate a Tl-Signature header value using a private key.
 */
final public class Signer {

    private final String kid;
    private final ECPrivateKey ecPrivateKey;

    private String path = "";
    private String method = "POST";
    private byte[] body = new byte[0];
    private final Map<HeaderName, String> headers = new LinkedHashMap<>();

    private Signer(String kid, ECPrivateKey ecPrivateKey) {
        this.kid = kid;
        this.ecPrivateKey = ecPrivateKey;
    }

    /**
     * Add the request method, defaults to `"POST"` if unspecified.
     *
     * @param method - the request method must be non null
     * @return the Signer instance
     * @throws IllegalArgumentException if the provided param is null
     */
    public Signer method(String method) {
        if (method == null)
            throw new IllegalArgumentException("the method must not be null");
        this.method = method;
        return this;
    }

    /**
     * Add the request absolute path starting with a leading `/` and without any trailing slashes.
     *
     * @param path - the request absolute path must not be null
     * @return the Signer instance
     * @throws IllegalArgumentException if the provided param is null or invalid
     */
    public Signer path(String path) {
        if (path == null)
            throw new IllegalArgumentException("the path must not be null");
        if(!path.startsWith("/"))
            throw new IllegalArgumentException("invalid path " + path + " must start with '/'");
        this.path = path;
        return this;
    }

    /**
     * Add the full request body. Note: This *must* be identical to what is sent with the request.
     *
     * @param body - the full request body must not be null
     * @return the Signer instance
     * @throws IllegalArgumentException if the provided param is null
     */
    public Signer body(byte[] body) {
        if (body == null)
            throw new IllegalArgumentException("the body must not be null");

        this.body = body;
        return this;
    }

    /**
     * Add the full request body. Note: This *must* be identical to what is sent with the request.
     *
     * @param body - the full request body must not be null
     * @return the Signer instance
     * @throws IllegalArgumentException if the provided param is null
     */
    public Signer body(String body) {
        if (body == null)
            throw new IllegalArgumentException("the body must not be null");

        this.body = body.getBytes(StandardCharsets.UTF_8);
        return this;
    }

    /**
     * Add a header name and value. May be called multiple times to add multiple different headers.
     * Warning: Only a single value per header name is supported.
     *
     * @param name  - must not be null
     * @param value - must not be null
     * @return the Signer instance
     * @throws IllegalArgumentException if the provided params are null
     */
    public Signer header(String name, String value) {
        if (name == null || value == null)
            throw new IllegalArgumentException("header name and value must not be null");
        this.headers.put(new HeaderName(name), value);
        return this;
    }

    /**
     * Add a Map of headers.
     * Warning: Only a single value per header name is supported.
     *
     * @param headers  - must not be null
     * @return the Signer instance
     * @throws IllegalArgumentException if the provided params are null
     */
    public Signer headers(Map<String, String> headers) {
        if (headers == null)
            throw new IllegalArgumentException("headers must not be null");
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            this.headers.put(new HeaderName(entry.getKey()), entry.getValue());
        }
        return this;
    }

    /**
     * Start building a request Tl-Signature header value using private key
     * RFC 7468 PEM-encoded data and the key's kid.
     *
     * @param kid           key identifier of the private key - must not be null
     * @param privateKeyPem the privateKey RFC 7468 PEM-encoded - must not be null
     * @return the Signer instance
     * @throws KeyException             if the provided key is invalid
     * @throws IllegalArgumentException if the provided params are null
     */
    public static Signer from(String kid, byte[] privateKeyPem) {
        if (kid == null || privateKeyPem == null)
            throw new IllegalArgumentException("kid and privateKey must not be null");

        ECPrivateKey privateKey = KeyException.evaluate(() ->
                ECKey.parseFromPEMEncodedObjects(new String(privateKeyPem)).toECKey().toECPrivateKey());

        return new Signer(kid, privateKey);
    }

    /**
     * Start building a request Tl-Signature header value using private key
     * RFC 7468 PEM-encoded data and the key's kid.
     *
     * @param kid           key identifier of the private key - must not be null
     * @param privateKeyPem the privateKey RFC 7468 PEM-encoded - must not be null
     * @return the Signer instance
     * @throws KeyException             if the provided key is invalid
     * @throws IllegalArgumentException if the provided params are null
     */
    public static Signer from(String kid, String privateKeyPem) {
        if (kid == null || privateKeyPem == null)
            throw new IllegalArgumentException("kid and privateKey must not be null");

        ECPrivateKey privateKey = KeyException.evaluate(() ->
                ECKey.parseFromPEMEncodedObjects(privateKeyPem).toECKey().toECPrivateKey());

        return new Signer(kid, privateKey);
    }

    /**
     * Start building a request Tl-Signature header value using private key
     * RFC 7468 PEM-encoded data and the key's kid.
     *
     * @param kid           key identifier of the private key - must not be null
     * @param privateKeyPem the privateKey RFC 7468 PEM-encoded - must not be null
     * @return the Signer instance
     * @throws KeyException             if the provided key is invalid
     * @throws IllegalArgumentException if the provided params are null
     */
    public static Signer from(String kid, ECPrivateKey privateKeyPem) {
        if (kid == null || privateKeyPem == null)
            throw new IllegalArgumentException("kid and privateKey must not be null");

        return new Signer(kid, privateKeyPem);
    }

    /**
     * Produce a JWS `Tl-Signature` v2 header value
     *
     * @return a JWS `Tl-Signature` v2 header value
     * @throws SignatureException if signature fails
     */
    public String sign() {
        return SignatureException.evaluate(() -> {
                    JWSHeader jwsHeader = JWSHeader.parse(jwsHeaderMap(this.kid, this.headers));

                    JWSObject jwsObject = new JWSObject(
                            jwsHeader,
                            new Payload(Utils.buildPayload(headers, method, path, body))
                    );

                    jwsObject.sign(new ECDSASigner(this.ecPrivateKey));

                    return jwsObject.serialize(true);
                }
        );
    }
}
